Migrated `app.config.chain.getById()` calls to newer `trpc.community.getCommunity()` + cleanup/state-refactor of related components

mzparacha opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/8762

## Description of Changes
- Removed all `app.config.chains.getById` references in favor of newer `trpc.community.getCommunity` route
- Refactored multiple frontend modules to cleanup/abstract/simplify logic relating to community state.
- Refactored frontend chain init according to `/bulkOffchain` API merged in `trpc.community.getCommunity` route
- Reduced usage of `ChainInfo` object usage

## "How We Fixed It"
N/A

## Test Plan
We have to ensure nothing that referenced `app.config.chain.getById()` broke
- Test all community pages including gated/restricted/special-case pages/actions like snapshots/contests/admin-pages/token-gating/stakes for any regressions
- Test non-community pages specially profile/feed/notifications page for any regressions
- Test site admin page for any regressions

## Deployment Plan
N/A

## Other Considerations
The changes in this pr will lead to some more cleanup/refinement. Created these tickets to make those changes as a follow-up.
- https://github.com/hicommonwealth/commonwealth/issues/8806
- https://github.com/hicommonwealth/commonwealth/issues/8811
- https://github.com/hicommonwealth/commonwealth/issues/8812

mzparacha commented:
Some CI issues for the `any` type still exist, will be fixing those shortly. For some strange reason I am not getting these lint errors when running `pnpm check-types` in my local env.

masvelio commented on packages/commonwealth/client/scripts/hooks/useRunOnceOnCondition.ts:
```diff
@@ -0,0 +1,25 @@
+  shouldRun: boolean;
+};
+
+const useRunOnceOnCondition = ({
```
what usecase made you to create and use this hook?

mzparacha replied:
Sometimes we have to set the component state once, after a react query call finishes. For this, we have been using the `useRef` flag approach. ex:

```
const data = useReactQuery();

useEffect(() => {
    if (!isDataSet.current && data) {   // set component data....   }
},[data])
```

this was getting duplicated in multiple places, so I created the hook which would be simpler to manage

ex:
```
 useRunOnceOnCondition({
    callback: () => {
        // the normal callback we do inside a useEffect
      );
    },
    shouldRun: the_condition_on_which_the_hook_should_run_once,
  });
```

masvelio commented on packages/commonwealth/client/scripts/models/IChainAdapter.ts:
```diff
@@ -44,37 +40,20 @@ abstract class IChainAdapter<C extends Coin, A extends Account> {
-      }),
-    ]);
 
+    // TODO: Do we really need this for 1inch only? -- ask product.
```
as far as I remember it was a request for 1inch as they wanted to have dark mode by default. It is done by not elegant way though. Worth to double check

masvelio commented on packages/commonwealth/client/scripts/state/api/communities/getCommuityById.ts:
```diff
@@ -0,0 +1,92 @@
+
+  // HACK: with @trpc/react-query v10.x, we can't directly call an endpoint outside of 'react-context'
+  // with this way the api can be used in non-react files. This should be cleaned up when we migrate
+  // to @trpc/react-query v11.x
```
can you create ticket for this migration

mzparacha replied:
@trpc/react-query v11.x is still with the @next flag (active development + can have breaking changes) -- not sure if migrating now would be a good idea.

Will create a ticket to clean this up when we eventually migrate to @trpc/react-query v11.x

masvelio commented on packages/commonwealth/client/scripts/views/components/sidebar/external_links_module.tsx:
```diff
@@ -1,15 +1,30 @@
+    elements,
+    githubs,
+    remainingLinks,
+    // slacks, TODO: maybe add this here as well -- ask product?
```
ask and implement or remove comment

mzparacha replied:
per product: we should display it like a normal link for now

masvelio commented on packages/commonwealth/client/scripts/views/modals/ManageCommunityStakeModal/ManageCommunityStakeModal.tsx:
```diff
@@ -13,22 +13,27 @@ import {
 } from './types';
 
+import { ChainBase } from '@hicommonwealth/shared';
+import { useManageCommunityStakeModalStore } from 'client/scripts/state/ui/modals';
```
can you skip `client/scripts/`

masvelio reviewed (commented):
half way through 😓 
now going to do some manual testing

masvelio reviewed (commented):
1. getting this error on dashboard page load

```
[13:22:00.257] ERROR (index.ts): INTERNAL_SERVER_ERROR: [ZodError] subscription.getCommentSubscriptions: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [
      0,
      "Comment",
      "Thread"
    ],
    "message": "Expected object, received null"
  }
] 
    err: {
      "type": "ZodError",
      "message": "[\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"object\",\n    \"received\": \"null\",\n    \"path\": [\n      0,\n      \"Comment\",\n      \"Thread\"\n    ],\n    \"message\": \"Expected object, received null\"\n  }\n]",
      "stack":
          ZodError: [
            {
              "code": "invalid_type",
              "expected": "object",
              "received": "null",
              "path": [
                0,
                "Comment",
                "Thread"
              ],
              "message": "Expected object, received null"
            }
          ]
              at get error [as error] (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/zod@3.23.6/node_modules/zod/lib/index.mjs:587:31)
              at ZodArray.parseAsync (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/zod@3.23.6/node_modules/zod/lib/index.mjs:715:22)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async outputMiddleware (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/index.mjs:308:26)
              at async callRecursive (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/index.mjs:451:32)
              at async callRecursive (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/index.mjs:451:32)
              at async resolve (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/index.mjs:481:24)
              at async inputToProcedureCall (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/resolveHTTPResponse-2fc435bb.mjs:46:22)
              at async Promise.all (index 0)
              at async resolveHTTPResponse (file:///Users/marcin/Desktop/Projects/commonwealth/node_modules/.pnpm/@trpc+server@10.45.2/node_modules/@trpc/server/dist/resolveHTTPResponse-2fc435bb.mjs:182:37)
      "aggregateErrors": [
        {
          "type": "Object",
          "message": "Expected object, received null",
          "stack":
              
          "code": "invalid_type",
          "expected": "object",
          "received": "null",
          "path": [
            0,
            "Comment",
            "Thread"
          ]
        }
      ],
      "issues": [
        {
          "code": "invalid_type",
          "expected": "object",
          "received": "null",
          "path": [
            0,
            "Comment",
            "Thread"
          ],
          "message": "Expected object, received null"
        }
      ],
      "name": "ZodError"
    }
```

2. this api cal looks weird (maybe it is ok but just want to double check)
![image](https://github.com/user-attachments/assets/ec250637-ff46-4f35-8465-ccc15ff3f9eb)

3. Filtering by stake does not work 

![image](https://github.com/user-attachments/assets/024f0f10-b58d-4349-a508-cb9de578da3a)

4. getting stake error on community load 
![image](https://github.com/user-attachments/assets/f3ff0914-0f06-4961-b986-d26dde0b8931)
Actually nothing related to stake does not work, I can't see indicator in left sidebar

5. infinite loader on create community flow when reserving namespace

![image](https://github.com/user-attachments/assets/c2324e71-7947-48fd-bcdf-06cd9bb3f888)

6. community I just created -> I am not admin of this community, cannot see admin capabilities section 

![image](https://github.com/user-attachments/assets/d9106a8c-858d-4f0c-8956-4c8d13a2ea51)

7. search throws backend errors
![image](https://github.com/user-attachments/assets/76af29f2-4b13-471e-86de-ebc2470110a8)

mzparacha commented:
Re: https://github.com/hicommonwealth/commonwealth/pull/8779#pullrequestreview-2235322544
1. This is on the `master` branch as well -- this is some "types" issue on the API side.
2. normal - trpc batches multiple requests (if different calls are made in a short timespan) -- the caches of responses are managed normally <img width="814" alt="image" src="https://github.com/user-attachments/assets/af3d505d-7f0b-4e56-878c-b3b38ac26525">
3. TODO
4. TODO
5. TODO
6. Fixed in https://github.com/hicommonwealth/commonwealth/pull/8779/commits/6b3889a438d774f596cd649ffe892f9c8b337d8b
7. This seems to be a "types" issue with the get community route. Basically the `community.network` is typed with `ChainNetwork`, and `ChainNetwork` doesn't include all the networks that communities in db might have. Ex: `Regen` is not listed in `ChainNetwork` but exists in the Communities table, visiting the `regen` community would break the route `http://localhost:8080/api/v1/community.getCommunity?batch=1&input=%7B%220%22%3A%7B%22id%22%3A%22regen%22%2C%22include_node_info%22%3Atrue%7D%7D`. I found ~59 more networks in my local db not listed in `ChainNetwork` and suspect there are probably more in prod db. We should list all those in `ChainNetwork` or maybe change/remove the type.

mzparacha commented:
@masvelio Thanks for the detailed review. I addressed all the feedback in https://github.com/hicommonwealth/commonwealth/pull/8779#issuecomment-2286614103

Also cc: @jnaviask - point `7` in https://github.com/hicommonwealth/commonwealth/pull/8779#issuecomment-2286614103 relates to the chain network issues we discussed.

masvelio commented on packages/commonwealth/client/scripts/views/pages/DirectoryPage/CommunityDirectoryCard/index.tsx:
can you avoid index.tsx for component definition, please 🙏

masvelio commented on packages/commonwealth/client/scripts/views/pages/NotificationSettingsOld/SubscriptionEntry.tsx:
```diff
@@ -0,0 +1,76 @@
+  onToggleReceiveInAppNotifications: () => void;
+};
+
+const SubscriptionEntry = ({
```
good call for this but let's try to avoid additional refactoring in the future in such a big PRs

masvelio reviewed (approved):
1. Zod validation error on communities page


https://github.com/user-attachments/assets/7b1f04ad-4368-41f7-a63c-203756cd3286

Tested issues I found yesterday, work well now, thanks for fixing them. Added couple small comments + found 1 issue with zod validation. Apart from that, looks good to me. Fantastic job!

mzparacha commented:
Re: https://github.com/hicommonwealth/commonwealth/pull/8779#pullrequestreview-2237504063

> 1. Zod validation error on communities page

This also relates to `7` in https://github.com/hicommonwealth/commonwealth/pull/8779#issuecomment-2286614103

cc: @masvelio @jnaviask

mzparacha commented:
The `7` from https://github.com/hicommonwealth/commonwealth/pull/8779#issuecomment-2286614103 was fixed in https://github.com/hicommonwealth/commonwealth/pull/8843 -- Thanks to @jnaviask

jnaviask commented:
Member count seems broken on the /members page but otherwise this PR seems good to me.

jnaviask reviewed (approved):
Members issue looks like it's my machine and not part of this PR.
